pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+
+// AttacherConfig defines the configuration for the attacher
+type AttacherConfig struct {
+	// Rules defines a series of rules that tell the attacher how to attach the probes
+	Rules []*AttachRule
+
+	// ScanProcessesInterval defines the interval at which we scan for terminated processes and new processes we haven't seen
+	ScanProcessesInterval time.Duration
+
+	// EnablePeriodicScanNewProcesses defines whether the attacher should scan for new processes periodically (with ScanProcessesInterval)
+	EnablePeriodicScanNewProcesses bool
+
+	// ProcRoot is the root directory of the proc filesystem
+	ProcRoot string
+
+	// ExcludeTargets defines the targets that should be excluded from the attacher
+	ExcludeTargets ExcludeMode
+
+	// EbpfConfig is the configuration for the eBPF program
+	EbpfConfig *ebpf.Config
+
+	// PerformInitialScan defines if the attacher should perform an initial scan of the processes before starting the monitor
+	PerformInitialScan bool
+
+	// ProcessMonitorEventStream defines whether the process monitor is using the event stream
+	ProcessMonitorEventStream bool
+
+	// EnableDetailedLogging makes the attacher log why it's attaching or not attaching to a process
+	// This is useful for debugging purposes, do not enable in production.
+	EnableDetailedLogging bool
+}
+
+// SetDefaults configures the AttacherConfig with default values for those fields for which the compiler
+// defaults are not enough
+func (ac *AttacherConfig) SetDefaults() {
+	if ac.ScanProcessesInterval == 0 {
+		ac.ScanProcessesInterval = 30 * time.Second
+	}
+
+	if ac.ProcRoot == "" {
+		ac.ProcRoot = "/proc"
+	}
+
+	if ac.EbpfConfig == nil {
+		ac.EbpfConfig = ebpf.NewConfig()
+	}
+}
+
+// Validate checks whether the configuration is valid, returns nil if it is, an error message otherwise
+func (ac *AttacherConfig) Validate() error {
+	var errs []string
+
+	if ac.EbpfConfig == nil {
+		errs = append(errs, "missing ebpf config")
+	}
+
+	if ac.ProcRoot == "" {
+		errs = append(errs, "missing proc root")
+	}
+
+	for _, rule := range ac.Rules {
+		err := rule.Validate()
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+
+	return errors.New("invalid attacher configuration: " + strings.Join(errs, ", "))
+}
+
+// ProbeManager is an interface that defines the methods that a Manager implements,
+// so that we can replace it in tests for a mock object
+type ProbeManager interface {
+	AddHook(string, *manager.Probe) error
+	DetachHook(manager.ProbeIdentificationPair) error
+	GetProbe(manager.ProbeIdentificationPair) (*manager.Probe, bool)
+}
+
+// FileRegistry is an interface that defines the methods that a FileRegistry implements, so that we can replace it in tests for a mock object
+type FileRegistry interface {
+	Register(namespacedPath string, pid uint32, activationCB, deactivationCB, alreadyRegistered utils.Callback) error
+	Unregister(uint32) error
+	Clear()
+	GetRegisteredProcesses() map[uint32]struct{}
```

Comment:
I think the interface methods should contain the documentation of the signatures